src/library_wasmfs.js

```diff
@@ -392,7 +398,61 @@ FS.createPreloadedFile = FS_createPreloadedFile;
         return __wasmfs_mknod(pathBuffer, mode, dev);
       }));
     },
-    // TODO: mkdev
+    makedev: (ma, mi) => ((ma) << 8 | (mi)),
+    registerDevice(dev, ops) {
+      var backendPointer = _wasmfs_create_jsimpl_backend();
+      var definedOps = {
+        userRead: ops.read,
+        userWrite: ops.write,
+
+        allocFile: (file) => {
+          wasmFSDeviceStreams[file] = {}
+        },
+        freeFile: (file) => {
+          wasmFSDeviceStreams[file] = undefined;
+        },
+        getSize: (file) => {},
+        read: (file, buffer, length, offset) => {
+          var bufferArray = Module.HEAP8.subarray(buffer, buffer + length);
+          try {
+            var bytesRead = definedOps.userRead(wasmFSDeviceStreams[file], bufferArray, 0, length, offset);
+          } catch (e) {
+            return -e.errno;
```

Comment:
Interesting, do these functions throw exceptions with an errno? Is there an example of that in the test suite or docs?

Reply:
I did see a test that simulated a "broken device" in `test_unistd_io`, and I think the try/catch is matching the expected behavior here:
https://github.com/emscripten-core/emscripten/blob/8ab7b6d34a53113551fd211c0dd1ea37d2bac2d3/test/unistd/io.c#L53-L61

Reply:
I see, thanks. I guess that API has made sense for JS all along, so good to match it here.